File: Optima/Math/CanonicalMatrix.cpp
// Optima is a C++ library for numerical solution of linear and nonlinear programing problems.
//
// Copyright (C) 2014-2016 Allan Leal
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see <http://www.gnu.org/licenses/>.

#include "CanonicalMatrix.hpp"

// Eigen includes
#include <Optima/Common/Exception.hpp>
#include <Optima/Math/Eigen/LU>

namespace Optima {

CanonicalMatrix::CanonicalMatrix()
{}

CanonicalMatrix::CanonicalMatrix(const Matrix& A)
{
	compute(A);
}

auto CanonicalMatrix::S() const -> const Matrix&
{
	return m_S;
}

auto CanonicalMatrix::R() const -> const Matrix&
{
	return m_R;
}

auto CanonicalMatrix::Rinv() const -> const Matrix&
{
	return m_Rinv;
}

auto CanonicalMatrix::P() const -> const PermutationMatrix&
{
	return m_P;
}

auto CanonicalMatrix::Q() const -> const PermutationMatrix&
{
	return m_Q;
}

auto CanonicalMatrix::ili() const -> Indices
{
	PermutationMatrix Ptr = m_P.transpose();
	auto begin = Ptr.indices().data();
	return Indices(begin, begin + rows());
}

auto CanonicalMatrix::ibasic() const -> Indices
{
	auto begin = m_Q.indices().data();
	return Indices(begin, begin + rows());
}

auto CanonicalMatrix::inonbasic() const -> Indices
{
	auto begin = m_Q.indices().data();
	return Indices(begin + rows(), begin + cols());
}

auto CanonicalMatrix::compute(const Matrix& A) -> void
{
	// The number of rows and columns of A
	const Index m = A.rows();
	const Index n = A.cols();

	// Check if number of columns is greater/equal than number of rows
	Assert(n >= m, "Could not canonicalize the given matrix.",
		"The given matrix has more rows than columns.");

	// Compute the full-pivoting LU of A so that P*A*Q = L*U
	Eigen::FullPivLU<Matrix> lu(A);

	// Get the rank of matrix A
	const Index r = lu.rank();

	// Get the LU factors of matrix A
	const auto Lbb = lu.matrixLU().topLeftCorner(r, r).triangularView<Eigen::UnitLower>();
	const auto Ubb = lu.matrixLU().topLeftCorner(r, r).triangularView<Eigen::Upper>();
	const auto Ubn = lu.matrixLU().topRightCorner(r, n - r);

	// Set the permutation matrices P and Q
	m_P = lu.permutationP();
	m_Q = lu.permutationQ();

	// Calculate the regularizer matrix R
	m_R = m_P;
	m_R.conservativeResize(r, m);
	m_R = Lbb.solve(m_R);
	m_R = Ubb.solve(m_R);

	// Calculate the inverse of the regularizer matrix R
	m_Rinv = m_P.transpose();
	m_Rinv.conservativeResize(m, r);
	m_Rinv = m_Rinv * Lbb;
	m_Rinv = m_Rinv * Ubb;

	// Calculate matrix S
	m_S = Ubn;
	m_S = Ubb.solve(m_S);

	// Initialize the permutation matrices Kb and Kn
	m_Kb.setIdentity(r);
	m_Kn.setIdentity(n - r);
}

auto CanonicalMatrix::swap(Index ib, Index in) -> void
{
	// Auxiliary references
	auto& M = m_M;
	auto& Q = m_Q.indices();
	auto& S = m_S;
	auto& R = m_R;
	auto& Rinv = m_Rinv;

	// Check if S(ib, in) is different than zero
	Assert(S(ib, in), "Could not swap basic and non-basic components.",
		"Expecting a non-basic component with non-zero pivot.");

	// Initialize the matrix M
	M = S.col(in);

	// Auxiliary variables
	const Index m = S.rows();
	const double aux = 1.0/S(ib, in);

	// Updadte the canonicalizer matrix R
	R.row(ib) *= aux;
	for(Index i = 0; i < m; ++i)
		if(i != ib) R.row(i) -= S(i, in) * R.row(ib);

	// Updadte the inverse of the canonicalizer matrix R
	Rinv.col(ib) = Rinv * S.col(in);

	// Updadte matrix S
	S.row(ib) *= aux;
	for(Index i = 0; i < m; ++i)
		if(i != ib) S.row(i) -= S(i, in) * S.row(ib);
	S.col(in) = -M*aux;
	S(ib, in) = aux;

	// Update the permutation matrix Q
	std::swap(Q[ib], Q[m + in]);
}

auto CanonicalMatrix::update(const Vector& w) -> void
{
	// Auxiliary variables
	const Index m = rows();
	const Index n = cols();

	// Auxiliary references to member data
	auto& Q = m_Q.indices();
	auto& S = m_S;
	auto& R = m_R;
	auto& Rinv = m_Rinv;
	auto& Kb = m_Kb;
	auto& Kn = m_Kn;

	// The indices and weights of the non-basic components
	auto inonbasic = Q.bottomRows(n - m);
	auto wn = Optima::rows(w, inonbasic);

	// Swap basic and non-basic components when the latter has higher weight
	Index j;
	for(Index i = 0; i < m; ++i)
	{
		const double wi = std::abs(w[Q[i]]);
		const double wj = abs(wn % tr(S.row(i))).maxCoeff(&j);
		if(wi < wj)
			swap(i, j);
	}

	// Sort the basic components in descend order of weights
	std::sort(Kb.indices().data(), Kb.indices().data() + m,
		[&](Index l, Index r) { return std::abs(w[Q[l]]) > std::abs(w[Q[r]]); });

	// Sort the non-basic components in descend order of weights
	std::sort(Kn.indices().data(), Kn.indices().data() + n - m,
		[&](Index l, Index r) { return std::abs(w[Q[m + l]]) > std::abs(w[Q[m + r]]); });

	// Rearrange the rows of S based on the new order of basic components
	Kb.applyThisOnTheLeft(S);

	// Rearrange the columns of S based on the new order of non-basic components
	Kn.applyThisOnTheRight(S);

	// Rearrange the rows of R based on the new order of basic components
	Kb.applyThisOnTheLeft(R);

	// Rearrange the columns of inv(R) based on the new order of basic components
	Kb.transpose().applyThisOnTheRight(Rinv);

	// Rearrange the permutation matrix Q based on the new order of basic and non-basic components
	Q.topRows(m).noalias() = Kb * Q.topRows(m);
	Q.bottomRows(n - m).noalias() = Kn * Q.bottomRows(n - m);
}

} // namespace Optima

File: tests/CanonicalMatrix.py
# Optima is a C++ library for numerical solution of linear and nonlinear programing problems.
#
# Copyright (C) 2014-2018 Allan Leal
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see <http://www.gnu.org/licenses/>.

from optima import *
from numpy import *
from numpy.testing import assert_almost_equal
from pytest import mark, raises
from utils.matrices import *

tested_nx  = [10]  # The tested number of x variables
tested_np  = [0, 5]       # The tested number of p variables
tested_ny  = [2, 5]       # The tested number of y variables
tested_nz  = [0, 2]        # The tested number of z variables
tested_nl  = [0, 1]        # The tested number of linearly dependent rows in Ax
tested_nu  = [0, 1]        # The tested number of unstable variables

# tested_nx  = [5, 10, 20, 50]  # The tested number of x variables
# tested_np  = [0, 5, 10]       # The tested number of p variables
# tested_ny  = [2, 5, 10]       # The tested number of y variables
# tested_nz  = [0, 2, 5]        # The tested number of z variables
# tested_nl  = [0, 1, 2]        # The tested number of linearly dependent rows in Ax
# tested_nu  = [0, 1, 2]        # The tested number of unstable variables

@mark.parametrize("nx", tested_nx)
@mark.parametrize("np", tested_np)
@mark.parametrize("ny", tested_ny)
@mark.parametrize("nz", tested_nz)
@mark.parametrize("nl", tested_nl)
@mark.parametrize("nu", tested_nu)
def testCanonicalMatrix(nx, np, ny, nz, nl, nu):

    basedims = BaseDims(nx, np, ny, nz)

    set_printoptions(linewidth=10000)

    # Ensure nx is larger than np and (ny + nz)
    if nx < np or nx < ny + nz: return

    # Ensure nl < ny
    if ny <= nl: return

    # # H = createMatrixViewHxxHxp(nx, np)
    # H = createMatrixViewH(nx, np)
    # V = createMatrixViewV(nx, np)
    # W = createMatrixViewW(nx, np, ny, nz, nl)

    # Wbar = createMatrixViewWbar(W)
    # js, ju = createStableUnstableIndices(nu, Wbar)

    # ju = Wbar.jn[:nu]  # the first nu` non-basic variables are considered unstable
    # js = array(set(arange(nx)) - set(ju))

    M = createMasterMatrix(nx, np, ny, nz, nl, nu)

    H = M.H
    V = M.V
    W = M.W

    canonicalmatrix = CanonicalMatrix(basedims)
    canonicalmatrix.update(M)

    Mbar = canonicalmatrix.view()

    #==========================================================================
    # Check jb, jn, js, ju in the canonical form of master matrix
    #==========================================================================

    # Assert canonical forms of W and M have the same basic and non-basic variables
    assert set(Mbar.jb) == set(M.RWQ.jb)
    assert set(Mbar.jn) == set(M.RWQ.jn)

    # Assert the unstable variables are non-basic only
    assert set(Mbar.ju).issubset(set(M.RWQ.jn))

    # Assert js and ju are disjoint and its union comprehends all variables
    assert set(Mbar.js) & set(Mbar.ju) == set()
    assert set(Mbar.js) | set(Mbar.ju) == set(range(nx))

    # Assert js and ju are disjoint and its union comprehends all variables
    assert set(Mbar.js) == set(M.js)
    assert set(Mbar.ju) == set(M.ju)

    #==========================================================================
    # Check Hss, Hsp, Vps, Vpp in the canonical form of master matrix
    #==========================================================================
    js = Mbar.js
    ju = Mbar.ju

    print(f"js = \n{js}")
    print(f"Mbar.Hss = \n{Mbar.Hss}")
    print(f"M.H.Hxx[:, js][js, :] = \n{M.H.Hxx[:, js][js, :]}")
    print(f"M.H.Hxx = \n{M.H.Hxx}")
    assert all(Mbar.Hss == M.H.Hxx[:, js][js, :])
    assert all(Mbar.Hsp == M.H.Hxp[js, :])

    assert all(Mbar.Vps == M.V.Vpx[:, js])
    assert all(Mbar.Vpp == M.V.Vpp)

    #==========================================================================
    # Check Sbn, Sbp, R in the canonical form of master matrix
    #--------------------------------------------------------------------------
    # Tests below are based on the fact that Rb*W*Q = [Ibb Sbn Sbp] where
    # Q = (jb, jn) and R = [Rb; Rl], Rl is associated with linearly
    # dependent rows in Wx.
    #==========================================================================
    nb  = Mbar.dims.nb
    nbs = Mbar.dims.nbs
    nns = Mbar.dims.nns

    jb  = Mbar.jb
    jn  = Mbar.jn
    jbs = jb[:nbs]
    jns = jn[:nns]

    Rbs = Mbar.Rbs

    Ibsbs = eye(nbs)
    Sbsns = Mbar.Sbsns
    Sbsp  = Mbar.Sbsp

    assert_almost_equal(Ibsbs, Rbs @ M.W.Wx[:, jbs])
    assert_almost_equal(Sbsns, Rbs @ M.W.Wx[:, jns])
    assert_almost_equal(Sbsp,  Rbs @ M.W.Wp)

    #==========================================================================
    # Check dimension variables in the canonical form of master matrix
    #==========================================================================

    # Compute expected dimension variables
    nw = ny + nz

    nu = len(ju)
    ns = nx - nu

    nb = len(jb)
    nn = len(jn)
    nl = nw - nb

    nbs = len(set(jb) - set(ju))
    nbu = len(set(jb) & set(ju))
    nns = len(set(jn) - set(ju))
    nnu = len(set(jn) & set(ju))

    jbs = jb[:nbs]
    jns = jn[:nns]

    nbe = len([i for i in range(nbs) if abs(H.Hxx[jbs[i], jbs[i]]) >= max(1.0, abs(V.Vpx[:, jbs[i]]).max(initial=0.0))])
    nne = len([i for i in range(nns) if abs(H.Hxx[jns[i], jns[i]]) >= max(abs(Sbsns[:, i]).max(initial=0.0), abs(V.Vpx[:, jns[i]]).max(initial=0.0))])

    nbi = nbs - nbe
    nni = nns - nne

    dims = Mbar.dims

    assert dims.nx == nx
    assert dims.np == np
    assert dims.ny == ny
    assert dims.nz == nz
    assert dims.nw == nw
    assert dims.ns == ns
    assert dims.nu == nu
    assert dims.nb == nb
    assert dims.nn == nn
    assert dims.nl == nl
    assert dims.nbs == nbs
    assert dims.nbu == nbu
    assert dims.nns == nns
    assert dims.nnu == nnu
    assert dims.nbe == nbe
    assert dims.nbi == nbi
    assert dims.nne == nne
    assert dims.nni == nni

    # ==========================================================================
    # Check method MasterMatrix::matrix()
    # ==========================================================================

    # Hxx = zeros((nx, nx))
    # for i in range(ns):
    #     for j in range(ns):
    #         Hxx[js[i], js[j]] = Mbar.Hss[i, j]
    # Hxx[ju, ju] = 1.0

    # Hxp = zeros((nx, np))
    # Hxp[js, :] = Mbar.Hsp

    # WxT = zeros((nx, nw))
    # WxT[js, :] = Mbar.Ws.T

    # Vpx = zeros((np, nx))
    # Vpx[:, js] = Mbar.Vps

    # Vpp = Mbar.Vpp

    # Wx = zeros((nw, nx))
    # Wx[:, js] = Mbar.Ws

    # Wp = Mbar.Wp

    # Opw = zeros((np, nw))
    # Oww = zeros((nw, nw))

    # Mmat = block([
    #     [Hxx, Hxp, WxT],
    #     [Vpx, Vpp, Opw],
    #     [ Wx,  Wp, Oww],
    # ])

    # assert all(Mmat == M.array())
